File: services/ai-detector/src/result.py
"""The messages AID sends back to api on the result-analysis queue.

The wire format is the AID branch of
`shared/schemas/result.analysis.v1.schema.json`. api is the single writer of
analysis rows (D-80), so this message is everything a batch will know about
this submission — including its failures, because a job that dead-letters
leaves api waiting until D-82's 30-minute sweep for an answer AID already has.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

from src.job import Detection

RESULT_QUEUE = "result-analysis"
SERVICE = "ai-detector"
VERSION = "1.0"

STATUS_COMPLETED = "completed"
STATUS_ERROR = "error"
STATUS_NOT_APPLICABLE = "not_applicable"


def _timestamp(at: datetime | None = None) -> str:
    moment = (at or datetime.now(timezone.utc)).astimezone(timezone.utc)
    return moment.strftime("%Y-%m-%dT%H:%M:%SZ")


def _envelope(job: Detection, status: str, at: datetime | None) -> dict[str, Any]:
    return {
        "analysis_submission_id": job.analysis_submission_id,
        "service": SERVICE,
        "status": status,
        # The job's trace id, not a new one: it is what joins the two sides of
        # this submission in Loki (D-88).
        "trace_id": job.trace_id,
        "timestamp": _timestamp(at),
        "version": VERSION,
    }


def completed(job: Detection, probability: float, at: datetime | None = None) -> dict[str, Any]:
    """A scored submission."""
    if not 0.0 <= probability <= 1.0:
        raise ValueError(f"probability {probability!r} is outside 0..1")

    return {**_envelope(job, STATUS_COMPLETED, at), "probability": probability}


def not_applicable(job: Detection, at: datetime | None = None) -> dict[str, Any]:
    """A language AID does not serve. `probability` is null, which the schema
    requires for any status other than completed."""
    return {**_envelope(job, STATUS_NOT_APPLICABLE, at), "probability": None}


def failed(job: Detection, cause: BaseException | str, at: datetime | None = None) -> dict[str, Any]:
    """A submission AID could not score.

    api stores a completed result with a null probability as an error anyway
    (D5) — the column is NOT NULL and telling an instructor a detection
    completed with nothing behind it is worse than reporting the failure — so
    the two sides agree on this shape.
    """
    return {
        **_envelope(job, STATUS_ERROR, at),
        "probability": None,
        # Optional in the schema and for debugging only: api logs it and
        # never shows it to a user (§2.5).
        "error_message": str(cause)[:1000],
    }


def encode(message: dict[str, Any]) -> bytes:
    """Renders a message body."""
    return json.dumps(message).encode()

File: services/plagiarism-checker/internal/result/result.go
// Package result builds the messages SIM sends back to api on the
// result-analysis queue.
//
// The wire format is the SIM branch of
// `shared/schemas/result.analysis.v1.schema.json`. api is the single writer of
// analysis rows (D-80), so this message is the whole of what a batch will ever
// know about this language group — including, deliberately, its failures: a
// group SIM cannot compare is answered with status=error rather than
// dead-lettered, because api is waiting and a message that never arrives costs
// the batch D-82's full 30-minute timeout for an answer SIM already has.
package result

import (
	"encoding/json"
	"time"

	"github.com/magecode/plagiarism-checker/internal/dolos"
	"github.com/magecode/plagiarism-checker/internal/job"
)

// ResultQueue is where api listens for finished analysis work (D-79).
const ResultQueue = "result-analysis"

// Statuses this service reports. They describe whether SIM got through the
// work, not how similar anything turned out to be.
const (
	StatusCompleted = "completed"
	StatusError     = "error"
)

// timeFormat is the schema's date-time, in UTC.
const timeFormat = "2006-01-02T15:04:05Z"

// Pair is one comparison on the wire. The nullable metrics are pointers so an
// unmeasured value encodes as null rather than as zero — api's columns are
// nullable and the two mean different things.
type Pair struct {
	SubmissionAID   int64   `json:"submission_a_id"`
	SubmissionBID   int64   `json:"submission_b_id"`
	Similarity      float64 `json:"similarity"`
	LongestFragment *int    `json:"longest_fragment"`
	TotalOverlap    *int    `json:"total_overlap"`
	ARegions        *string `json:"a_regions"`
	BRegions        *string `json:"b_regions"`
}

// SubmissionStatus is one submission's outcome inside the group.
type SubmissionStatus struct {
	AnalysisSubmissionID int64  `json:"analysis_submission_id"`
	Status               string `json:"status"`
	// ErrorMessage is omitted unless Status is error: the schema marks it
	// optional and an empty string reads as an error with no detail.
	ErrorMessage string `json:"error_message,omitempty"`
}

// Analysis is `result.analysis` v1.0, SIM branch.
type Analysis struct {
	AnalysisProblemID  int64              `json:"analysis_problem_id"`
	Service            string             `json:"service"`
	Status             string             `json:"status"`
	ErrorMessage       string             `json:"error_message,omitempty"`
	Language           string             `json:"language"`
	LanguageGroupIndex int                `json:"language_group_index"`
	LanguageGroupTotal int                `json:"language_group_total"`
	Pairs              []Pair             `json:"pairs"`
	SubmissionStatuses []SubmissionStatus `json:"submission_statuses"`
	TraceID            string             `json:"trace_id"`
	Timestamp          string             `json:"timestamp"`
	Version            string             `json:"version"`
}

// Completed builds the message for a group SIM compared.
//
// failures maps an analysis_submission_id to the reason its source never
// reached the workspace; those submissions are reported as errors while the
// rest of the group keeps its comparisons. api (D4) ingests both halves of
// that from the one message.
func Completed(j job.Similarity, pairs []dolos.Pair, failures map[int64]error, at time.Time) Analysis {
	message := envelope(j, at)
	message.Status = StatusCompleted
	message.Pairs = wirePairs(pairs)
	message.SubmissionStatuses = statuses(j, failures)
	return message
}

// Failed builds the message for a group SIM could not compare at all.
func Failed(j job.Similarity, cause error, at time.Time) Analysis {
	message := envelope(j, at)
	message.Status = StatusError
	message.ErrorMessage = cause.Error()
	message.Pairs = []Pair{}

	// Every submission carries the group's failure: api derives each
	// submission's plagiarism_status from this list, and one left out stays
	// in_queue until the sweeper.
	message.SubmissionStatuses = make([]SubmissionStatus, 0, len(j.Submissions))
	for _, submission := range j.Submissions {
		message.SubmissionStatuses = append(message.SubmissionStatuses, SubmissionStatus{
			AnalysisSubmissionID: submission.AnalysisSubmissionID,
			Status:               StatusError,
			ErrorMessage:         cause.Error(),
		})
	}

	return message
}

// envelope fills the fields both outcomes share. The group index and total are
// echoed rather than recomputed: api's completion set (U-9) is keyed on them,
// so a message that reports a different split can never complete its batch.
func envelope(j job.Similarity, at time.Time) Analysis {
	return Analysis{
		AnalysisProblemID:  j.AnalysisProblemID,
		Service:            job.QueueName,
		Language:           string(j.Language),
		LanguageGroupIndex: j.LanguageGroupIndex,
		LanguageGroupTotal: j.LanguageGroupTotal,
		Pairs:              []Pair{},
		SubmissionStatuses: []SubmissionStatus{},
		// The job's trace id, not a new one: it is what joins the two sides
		// of this batch in Loki (D-88).
		TraceID:   j.TraceID,
		Timestamp: at.UTC().Format(timeFormat),
		Version:   job.Version,
	}
}

func wirePairs(pairs []dolos.Pair) []Pair {
	wire := make([]Pair, 0, len(pairs))
	for _, pair := range pairs {
		wire = append(wire, Pair{
			SubmissionAID:   pair.SubmissionAID,
			SubmissionBID:   pair.SubmissionBID,
			Similarity:      pair.Similarity,
			LongestFragment: pair.LongestFragment,
			TotalOverlap:    pair.TotalOverlap,
			ARegions:        pair.ARegions,
			BRegions:        pair.BRegions,
		})
	}
	return wire
}

func statuses(j job.Similarity, failures map[int64]error) []SubmissionStatus {
	reported := make([]SubmissionStatus, 0, len(j.Submissions))
	for _, submission := range j.Submissions {
		status := SubmissionStatus{
			AnalysisSubmissionID: submission.AnalysisSubmissionID,
			Status:               StatusCompleted,
		}
		if cause, failed := failures[submission.AnalysisSubmissionID]; failed && cause != nil {
			status.Status = StatusError
			status.ErrorMessage = cause.Error()
		}
		reported = append(reported, status)
	}
	return reported
}

// Queue is where the message goes.
func (a Analysis) Queue() string { return ResultQueue }

// TraceIdentifier is the header value the publisher sets (D-88).
func (a Analysis) TraceIdentifier() string { return a.TraceID }

// Encode renders the message body.
func (a Analysis) Encode() ([]byte, error) { return json.Marshal(a) }
